### ISSUE TITLE
Utf8JsonReader.GetDouble: Should it return (+/-) infinity, or throw?

### ISSUE THREAD

**Issue by bartonjs** [dotnet/corefx#33997]
If `Utf8JsonReader` encounters the numeric value `1e+100000002` then GetDouble is returning `double.Infinity`.

In a hallway discussion it was believed that this was bad, since the reader would read "non-infinity", return "infinity" and then the writer would reject it (since Infinity is verboten in the JSON spec).

**Comment by AlexejLiebenthal** [comment]
I think this is a really special case and **should throw** (`ArgumentOutOfRangeException`) on any numeric (`double`, `int`, `short`, ..) getter, if the value doesn't fit.
### Why
- No roundtrip, as you already mentioned, would be possible (JSON > c# object > JSON).
- If such a big number occurs, this could be some special case in some special context and could have a need for a special treatment.
- As I know most programming languages follow the IEEE754 standard. So if anyone wrote a bigger number than ``double.MaxValue`` into the JSON, they've done it probably on purpose.

##### TL;DR
JSON is a subset of JS. JS has a IEEE754 (binary64) based number format. So IEEE754 should fit in most cases. But IEEE also defines +/- infinity and NaN. And JSON disallows this values. So maybe the schema needs to look like this to support IEEE754:
```json
"ieee754Binary64Value": {
  "anyOf": [{
      "type": "number",
      "minimum": -1.79769313486232E+308,
      "maximum": 1.79769313486232E+308
    },
    {
      "enum": ["NaN", "PositiveInfinity", "NegativeInfinity"]
    }
  ]
}
```
Like this is per schema definition/contract absolutely clear, what the JSON should look like. And from now on any number bigger or lower than these in IEEE defined is a schema vialation.
But if no schema is provided, we need to think that any number out of real numbers must fit into the type in which we like to store the deserialized value in, so that a perfect roundtrip without information loss is possible. So I think that, if any JSON gives the consumer values out of range of `double` (or other numeric types), than this should have/has a reason. And the code must deal with it. And if ``double`` can't it must throw.

**Comment by bartonjs** [comment]
> JSON is a subset of JS.

It isn't, really, when it comes to numbers.  JSON Number is an arbitrary precision "BigRational"; and thus holds an infinity of infinities (Aleph-1) of values that cannot be represented by IEEE754 / ECMAScript (since ECMA-262 9.0 says Number is "primitive value corresponding to a double-precision 64-bit binary format IEEE 754-2008 value")

> should throw (ArgumentOutOfRangeException) on any numeric (double, int, short, ..) getter, if the value doesn't fit.

`FormatException` and/or `OverflowException` are the two acceptable exceptions here; based on the prior art of `int.Parse(string)`.  IIRC we landed on

```C#
public void GetInt32()
{
    if (!TryGetInt32(out int value))
    {
        throw new FormatException("probably words go here");
    }

    return value;
}
```

which is the same for "too big" and "had a decimal point" (or a scentific-notation `e`).

**Comment by tannergooding** [comment]
For `double` and `float`, the IEEE 754 specification defines that numbers larger than the largest representable value overflows to `+/-Infinity` and values smaller than the smallest representable double underflow to `+/-Zero`. Javascript itself, being an IEEE compliant language has the same behavior (e.g. `1.79769313486232E+309` returns `Infinity` -- this is easily tested in any modern browser). In other words, there is no concept of a "number" that is "too big" or "too small" only the concept of a "number" that is not in a valid format.

JSON, however, disallows the "number" token from being `NaN` or `Infinity` -- The ECMA 404 spec indicates:
> Numeric values that cannot be represented as sequences of digits (such as Infinity and NaN) are not permitted.

However, the spec also indicates:
> JSON is agnostic about the semantics of numbers. In any programming language, there can be a variety of number types of various capacities and complements, fixed or floating, binary or decimal. That can make interchange between different programming languages difficult. JSON instead offers only the representation of numbers that humans use: a sequence of digits. All programming languages know how to make sense of digit sequences even if they disagree on internal representations. That is enough to allow interchange.

I believe the most accurate behavior (that follows both the IEEE and JSON specs as closely as possible) is that not all `numbers` are guaranteed to roundtrip when read and then written back or there needs to be a mechanism to determine if parsing the `number` will never succeed. The writer already has to handle double/float that are `Infinity` or `NaN`, so the only downside is that roundtripping a value is never guaranteed. However, not doing this behavior means that such files are also effectively "unparseable", since you will never be able to succesfully read the `number` into anything the framework can currently represent. It is also already the case that numbers may not 100% roundtrip (a given numeric string, parsed as a float/double, will be converted to the nearest representable value; which may be different from the exact value given).

**Comment by tannergooding** [comment]
It is also likely worth mentioning that, if you use the built-in `JSON.Parse` function in Javascript:
```js
var text = '{ "value": 1.79769313486232E+309 }';
var obj = JSON.parse(text);
obj.value
```

The result is `Infinity`

**Comment by tannergooding** [comment]
It also likely worth mentioning that if you use the built-in `JSON.Stringify` function in Javascript:
```
var text = '{ "value": 1.79769313486232E+309 }';
var obj = JSON.parse(text);
JSON.stringify(obj)
```

The result is `{"value":null}`

**Comment by AlexejLiebenthal** [comment]
@bartonjs 
> It isn't, really, when it comes to numbers. JSON Number is an arbitrary precision "BigRational"; and thus holds an infinity of infinities (Aleph-1) of values that cannot be represented by IEEE754 / ECMAScript (since ECMA-262 9.0 says Number is "primitive value corresponding to a double-precision 64-bit binary format IEEE 754-2008 value")

I know. I only liked to point out the history. And that most of the use cases the value provided in a json will be a valid IEEE754 64 bit binary. But the producer of the json is responsible of providing a schema file (like I hacked in here as an example). Even if it's not everywhere common, and quit naive from me, it is quite mandatory. So if I like to deserialize the json, I know which contract, the property has. So the schema tells me, if I can safely rely on `double`.

>FormatException and/or OverflowException are the two acceptable exceptions here; based on the prior art of int.Parse(string). IIRC we landed on

My fault. Totally agree on that.

@tannergooding 
> JSON, however, disallows the "number" token from being `NaN` or `Infinity` -- The ECMA 404 spec indicates:

This is what I try to say. And pointed out, that the JSON creator is responsible of providing a json schema to demonstrate the range or if its IEEE or whatever.
> But IEEE also defines +/- infinity and NaN. And JSON disallows this values.

> It is also already the case that numbers may not 100% roundtrip (a given numeric string, parsed as a float/double, will be converted to the nearest representable value; which may be different from the exact value given).

That's true.

> The result is `Infinity`

Hmm. Interesting. Following this pattern it shouldn't throw. But it feels somehow natural to me, that if a type doesn't fit, that it should throw.

> The result is `{"value":null}`

If the schema expects `"value": "type": "number"`, this trip produced an incorrect json.

-----

If the range is specified, the consumer can be sure, when deserializing to `double` that it fits. It is still not 100% roundtrip, but maybe, if this matters (e.g. exchange rates) decimal might be the right choice for deserializing the json `number`.

Could this also being implemented as an JsonReaderOption, if it throws or not?

**Comment by tannergooding** [comment]
I think it can basically be summed to two cases:
1. We fail if `number` would overflow to `+/-Infinity`, in which case there are some JSON inputs that you will never be able to handle by default (the token would be reported as a `number`, but all attempts to parse the token into a supported .NET type fail)
2. We allow `number` to overflow to `+/-Infinity`, in which case we are IEEE compliant but there are some values which will never roundtrip and users have to explicitly handle `+/-Infinity` (something that they already have to do for other cases)

I would think that 2 is the more consistent and better approach overall.

**Comment by bartonjs** [comment]
The real consequence of option 2 is that we should get rid of TryGetDouble, since it can never return false.  (If the value (e.g.) had two decimal points in it then the reader already threw); which is why we need to define the behavior for "too large" numbers.

(And then presumably JsonDocument will match)

**Comment by ahsonkhan** [comment]
@ahsonkhan

**Comment by AlexejLiebenthal** [comment]
> (the token would be reported as a `number`, but all attempts to parse the token into a supported .NET type fail)

If it is a quite Big Number `BigInteger` would be possible. (Big `double` values (>1.0E18) are already kind of an integer number)
If the number is between `double.Epsilon` and `0.0` and as C# does not support a type like `BigDecimal` this only can be interpreted as Zero.

The consumer must have appropriate DTOs in which he can deserialize the most correct value out of the json. Better it would be, if it could be everytime 100% correct, but as the nature of `double` there are only 16-18 relevant digits.

**Comment by ahsonkhan** [comment]
> I would think that 2 is the more consistent and better approach overall.

I think we should explicitly reject such large values (and return false/throw) instead of `Infinity` so that the JSON behavior between netstandard and .NET Core 3.0 is consistent. Otherwise, the different Utf8Parser semantics across platforms bleeds into the semantics of the JSON library and now the consumers will see different behavior.

This got highlighted by the test failing on NETFX here: https://github.com/dotnet/corefx/pull/34604#discussion_r249218313

cc @KrzysztofCwalina

**Comment by JamesNK** [comment]
What does JSON.Parse do?

**Comment by tannergooding** [comment]
@JamesNK, I called out what `JSON.Parse` does here: https://github.com/dotnet/corefx/issues/33997#issuecomment-447140932, it returns `infinity` and `JSON.Stringify` (called out just below that comment) returns `{"value":null}`

**Comment by tannergooding** [comment]
> so that the JSON behavior between netstandard and .NET Core 3.0 is consistent

This could also be resolved by special-casing those values in .NET Standard and I would strongly push as the preference if we said that the behavior has to match. However, we are also already in a world where Full Framework vs .NET Core and even older .NET Core vs new .NET Core can and will have different behaviors. We also have a large number of cases where the old logic simply won't roundtrip some numbers under any conditions (not just these edge cases), so I think that should be a non-goal.

My strong vote is that, on .NET Core, we should push for IEEE compliance and we should simply call out that there may be some values which don't roundtrip. We already have this scenario for `PositiveInfinity`, `NegativeInfinity`, and `NaN` when serializing an existing floating-point value and we will have more types in similar situations moving forward (such as `binary16` and `binary128` which are the half-precision and quad-precision IEEE types or user-defined types such as `System.Numerics.Complex`, which requires the user of infinity/nan). Both the JSON functionality in javascript and the behavior in Newtonsoft.JSON are matching this as well.
